File: solution/76/solution.py
from collections import defaultdict
class Solution(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        result = ''
        char_t = defaultdict(int)
        char_s = defaultdict(int)
        for x in t:
            char_t[x] += 1
        begin = 0
        for i, v in enumerate(s):
            char_s[v] += 1
            while begin < i:
                if char_t[s[begin]] == 0:
                    begin += 1
                elif char_s[s[begin]] > char_t[s[begin]]:
                    char_s[s[begin]] -= 1
                    begin += 1
                else:
                    break
            if self.is_Window_Ok(char_s, char_t) and (result == '' or i-begin+1 < len(result)):
                result = s[begin: i+1]
        return result


    def is_Window_Ok(self, char_s, char_t):
        for k,v in char_t.items():
            if char_t[k] > char_s[k]:
                return False
        return True

s = Solution()
print s.minWindow("ab", 'a')

File: solution/76/solution.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int j = 0;
        string result = "";
        int map_s[128] = {0};
        int map_t[128] = {0};
        vector<int> vec_t;
        for(int i = 0;i < t.length();i++){
            map_t[t[i]]++;
        }
        for(int i=0;i<128;i++){
            if(map_t[i]>0)vec_t.push_back(i);
        }
        for(int i=0;i<s.length();i++){
            map_s[s[i]]++;
            while(j<i){
                if(map_t[s[j]]==0){
                    j++;
                }else if(map_t[s[j]]<map_s[s[j]]){
                    map_s[s[j]]--;
                    j++;
                }else{break;}
            }
            if(isWindowOk(map_s, map_t, vec_t)){
                int newlen = i-j+1;
                if(result==""||newlen<result.length()){
                    result = s.substr(j, newlen);
                }
            }
        }
        return result;
    }
    bool isWindowOk(int map_s[], int map_t[], vector<int> &vec_t){
        for (int i = 0; i < vec_t.size(); i++){
            if (map_s[vec_t[i]] < map_t[vec_t[i]]){
                return false;
            }
        }
        return true;
    }
};